### DISCUSSION TITLE
Sensitivity of design parameters.

### DISCUSSION THREAD

**Discussion by aminrbspace** [casadi/casadi#3566] [Q&A]
Hi. I am currently facing challenges with MPC tuning for a dynamic system with fixed references. Finding optimal Q and R parameters is so difficult Moreover, when I attempt to change the references, the need for retuning arises, leading to a time-consuming process. Despite trying methods such as sqrt(1/max(states)) and employing the PSO algorithm, I am not entirely satisfied with the results.  Any insights you can share would be greatly appreciated!

**Comment by brunoml5** [comment]
Hi!

Tuning controllers sometimes is more about art and creativity. It depends a lot on your process so it would be necessary more information on what you are trying to solve.

What do you mean by "optimal" Q and R? Do you have some kind of cost function different from the traditional squared error? Maybe go straight for Economic MPC where the stage cost is your economic function.

The problem for EMPC is stability. So, use a large horizon at first and take a look at the result. Then, add quadratic weights on the state and input variables that you feel needs help tracking your reference. This is kind of a "regularized" EMPC.

Take a look on the references below explaining the concept. I tried writing a little bit about this in my [master's dissertation](https://scholar.google.com/citations?view_op=view_citation&hl=pt-BR&user=hBDog70AAAAJ&citation_for_view=hBDog70AAAAJ:2osOgNQ5qMEC). Maybe it is easier to understand.

Also, take a look at this [TuneMPC](https://github.com/jdeschut/tunempc/) toolbox based on the work from Zanon. I never tested it, but it tunes a tracking MPC which provides similar performance than your EMPC.

For hyperparameter optimization, try [pymoo](https://pymoo.org/).  I've used it before with Casadi and it works great. It is easy to test different global optimization algorithms.

I hope it helps!

[1] M. Ellis, H. Durand, and P. D. Christofides, “A tutorial review of economic model predictive control methods,” J. Process Control, vol. 24, no. 8, pp. 1156–1178, Aug. 2014, doi: 10.1016/j.jprocont.2014.03.010.

[1] J. B. Rawlings, D. Angeli, and C. N. Bates, “Fundamentals of economic model predictive control,” in 2012 IEEE 51st IEEE Conference on Decision and Control (CDC), 2012, pp. 3851–3861, doi: 10.1109/CDC.2012.6425822.

[1] D. W. Griffith, V. M. Zavala, and L. T. Biegler, “Robustly stable economic NMPC for non-dissipative stage costs,” J. Process Control, vol. 57, pp. 116–126, 2017, doi: 10.1016/j.jprocont.2017.06.016.

**Comment by aminrbspace** [comment]
Dear Mr. Lima,

Thank you so much for your time. To summarize my situation, I'm developing
a controller for multiple quadrotors with different mass and moment of
inertias using CasADi. The MPC cost function is quadratic(e'Qe+u'Ru), and
initially, I set a large prediction horizon. However, I've encountered
significant sensitivity in choosing Q and R.

For instance, when using Q=[(1.5/100)^2;(1.5/100)^2;(0.45/10)^2]*eye(3) and
R=(1/1900)^2*eye(4), the controller works well for a reference like
[5,5,-20]. But when I change the reference to [5,5,-40] (which is not much
to the previous one), everything goes wrong(diverge), even with the same Q
and R. I've tried tuning Q and R incrementally based on the previous
values, but it's cumbersome to track the new reference. This unexpected
sensitivity to Q and R adjustments is puzzling to me. I read the mentioned
references. I appreciate your valuable comments.

**Reply by brunoml5** [reply]
I'm not familiar with quadrotor models so I don't know the states and inputs, but I'll try.

Have you tried using directly penalizing the rate of u instead of u-u_sp? Then, the cost in u will not depend on the reference and might give better results.

Is this problem comming from the nonlinearity of the model? To test this hypothesis, change the model of your system and your controller to a linear one. Then, test again your change of references.

Is the process becoming unstable because of constraints being activated? Check the constraints, specially on the input, maybe there is some pattern there.

I don't know how I can help more with limited knowledge on quadrotors and your specific problem. It appears it is not directly related to casadi (unless this does not happen with other tools).

**Comment by aminrbspace** [comment]
Hello, apologies for my delayed response.

The fully nonlinear model performs well with small parameters of model
values, exhibiting stability without significant sensitivity to Q and R for
varying references.

However, a persistent challenge lies in the tuning of Q and R for larger
model values, which significantly influences convergence. Even minor
parameter adjustments necessitate a thorough returning process.

Efforts have been made to enhance the objective function by incorporating
both input rate and inputs. Additionally, constraints have been
meticulously checked several times.

1) Another  concern is the solver time. Despite a 10-second simulation at
100 Hz, the process takes an unexpectedly long 30 minutes.

2) Changes in constraints significantly impact results, even though I
believe there is a solution within a much narrower range of constraint
values(using other control techniques).  Looking for tips to enhance
stability,

 especially with Q, R, and other constraints.

Thank you for your patience and valuable insights.

Best regards,
Amin

**Reply by brunoml5** [reply]
Hi!

> 1) Another  concern is the solver time. Despite a 10-second simulation at
> 100 Hz, the process takes an unexpectedly long 30 minutes.

In my experience, if your solver is taking too long, it means that you are not building your NLP properly with multiple-shooting, collocation, adequate constraints, stuff like that. Try checking on that.

I don't know how I can help you without looking at your model and your code. It appears that it is not a Casadi problem but your model and code. Can you provide a small script that replicates the problem you are trying to solve? Maybe I'll take a quick look on that.

Apart from this, maybe ask around in ResearchGate specifically about your tuning problem for quadrotors. Furthermore, there must be tons of scientific articles and blog post about control of quadrotors. Take a look on that.

**Comment by aminrbspace** [comment]
Hi,

Thank you so much for your reply, especially about solver time. So, I will
make a small script and send it to you.

Bests,

**Comment by aminrbspace** [comment]
Dear Bruno,

Hi. I appreciate to help me based on the following explanations:

Objective: I'm working on NMPC control with MHE for actuator fault
estimation, using equations attached. Variables for MHE are states and
faults, with noise on measurements, process noise is considered zero.

Issue: Successful state estimation, but struggling with fault estimation
despite extensive tuning.

Method: Utilized fault OCP formulation from attached PDF for MHE.

Results: State estimation is accurate; however, fault estimation remains
problematic in simulations.

PDF Details: PDF shows real-time linear change in actuator fault (Gamma)
with corresponding fault estimation results.

Request for Feedback: I I want to know your ideas on MHE settings alignment
with PDF formulation for fault estimation. (I am not sure of the
implementation setting for Gamma, but the setting of P as a parameter and
others is correctly implemented in the main code; there is no problem on
that side)

I Appreciate your help! If you need more information, please let me know.

Best regards,
